blog/_src/posts/2019-08-19-scoping-in-r.md

```diff
@@ -0,0 +1,491 @@
+\[[DOI](https://doi.org/10.1080/10618600.2000.10474895)\]\[[Available
+online](https://www.stat.auckland.ac.nz/~ihaka/downloads/lexical.pdf)\]
+
+
+### Examples
+
+#### Adding variable mappings at run time
+
+```r
+x <- 1
+f <- function() {
+  g <- function() x
+  x <- 2
+  g()
+}
+f() # 2
+```
+
+When `f` is called, it creates a function `g` that returns `x`, assigns `2` to
+`x`, and then calls `g`. When `g` is called, it looks up `x`. Since no mapping
+is found in `g`'s environment, it searches in the enclosing environment
+(`f`'s), and finds that `x` has value `2`. Therefore, `g` returns `2`.
+
+Note that the `x` on line 3 is resolved only when function `g` is called, not
+when it is defined. However, when `g` is defined, its environment has a
+reference to `f`'s environment. Therefore, as long as `x` is defined _before_
+`g` is called, the lookup will always succeed.
+
+Here's a second example:
+
+```r
+x <- 1
+f <- function(b) {
+  if (b)
+    x <- 2
+  x
+}
+f(TRUE)  # 2
+f(FALSE) # 1
+```
+
+`f` is a function that branches on its argument, `b`. If `b` evaluates to true,
+then the expression `x <- 2` is evaluated, and a mapping for `x` is created in
+`f`'s environment. Otherwise, no mapping is created.
+
+When we look up the value of `x` on line 5, R will first search the function's
+environment. If `b` evaluated to true, then R will find a value for `x`, which
+is `2`. Otherwise, R will search in the enclosing environment of `f`, and find
+that `x` is `1`.
+
+Both of these examples vaguely resemble dynamic scoping, in that `x` takes the
+value of the most recent assignment. However, this is not how R is
+implemented---as there is no traversal of the call stack---and it is not
+consistent with how R behaves in other examples.
+
+#### Function lookup
+
+```r
+f <- function(x) x
+g <- function(f) {
+  f(0) # not an error
+}
+g(42) # 0
+```
+
+R has slightly different lookup rules, if the variable is in function call
+position. Specifically, R will search the environment chain and skip
+non-function values.
+
+In this example, we call `g` with the argument `42`, which is not a function.
+Then, in the body of `g`, we call `f(0)` on line 3, which requires looking up
+`f`. Although there is an `f` in the environment of `g`, its value is `42`,
+which is not a function. R will then search the enclosing environment, where it
+finds the function defined on line 1. Therefore, the lookup on line 3 resolves
+to the function on line 1, so `f(0)` returns `0`.
+
+This feature exists because `c` is the built-in function that constructs
+vectors (in other words, one of the most commonly used functions in R), but it
+is also a commonly used argument name.
+
+#### Super assignment
+
+```r
+x <- 0
+f <- function() {
+  x <- 1
+  x <<- 2
+  x
+}
+f() # 1
+x   # 2
+```
+
+`<<-` is the "super assignment" operator. It skips the current environment and
+then searches the chain of enclosing environments until it finds a variable to
```

Comment:
does it need to be a top level variable?

(I imagine this example would give the same result
```
x <- 0
f <- function() {
  x <- 1
  g <- function() {
    x <- 3
    x <<- 2
    x
  }
  g()
  x
}
f()
x 
```
)

Reply:
It'll skip the local variable and find the next one, so no, it doesn't have to be the top level. To annotate your example:

```
x <- 0
f <- function() {
  x <- 1
  g <- function() {
    x <- 3
    x <<- 2
    x # 3
  }
  g() # 3
  x # 2
}
f() # 2
x  # 0
```

And here's another example:
```
f <- function() {
  x <- 1
  x <<- 2
  x # 1
}
f() # 1
x # 2
```